src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -1358,69 +1356,65 @@ private void StartTerminatingConnection(int lastValidStream, Exception abortExce
-                    _lastValidStreamId = lastValidStream;
+                    _lastStreamId = lastValidStream;
                 }
                 else
                 {
                     // We have already received GOAWAY before
                     // In this case the smaller valid stream is used
-                    _lastValidStreamId = Math.Min(_lastValidStreamId, lastValidStream);
+                    _lastStreamId = Math.Min(_lastStreamId, lastValidStream);
                 }
 
-                if (NetEventSource.IsEnabled) Trace($"{nameof(lastValidStream)}={lastValidStream}, {nameof(_lastValidStreamId)}={_lastValidStreamId}");
-
-                bool hasAnyActiveStream = false;
+                if (NetEventSource.IsEnabled) Trace($"{nameof(lastValidStream)}={lastValidStream}, {nameof(_lastStreamId)}={_lastStreamId}");
 
                 foreach (KeyValuePair<int, Http2Stream> kvp in _httpStreams)
                 {
                     int streamId = kvp.Key;
                     Debug.Assert(streamId == kvp.Value.StreamId);
 
-                    if (streamId > _lastValidStreamId)
+                    if (streamId > _lastStreamId)
                     {
-                        kvp.Value.OnAbort(abortException, canRetry: true);
-                        _httpStreams.Remove(streamId);
+                        streamsToAbort.Add(kvp.Value);
                     }
                     else
                     {
-                        if (NetEventSource.IsEnabled) Trace($"Found {nameof(streamId)} {streamId} <= {_lastValidStreamId}.");
-                        hasAnyActiveStream = true;
+                        if (NetEventSource.IsEnabled) Trace($"Found {nameof(streamId)} {streamId} <= {_lastStreamId}.");
                     }
                 }
 
-                if (!hasAnyActiveStream)
-                {
-                    _disposed = true;
-                }
-
                 CheckForShutdown();
             }
+
+            // Avoid calling OnAbort under the lock, as it may cause the Http2Stream
+            // to call back in to RemoveStream
+            foreach (Http2Stream s in streamsToAbort)
+            {
+                s.OnReset(abortException, canRetry: true);
+            }
         }
 
         private void CheckForShutdown()
         {
-            Debug.Assert(_disposed || _lastValidStreamId != -1);
+            Debug.Assert(_disposed || _lastStreamId != -1);
             Debug.Assert(Monitor.IsEntered(SyncObject));
 
             // Check if dictionary has become empty
             if (_httpStreams.Count != 0)
             {
                 return;
             }
-            else
-            {
-                _disposed = true;
-            }
 
             // Do shutdown.
             _stream.Close();
 
-            _connectionWindow.Dispose(_abortException);
-            _concurrentStreams.Dispose(_abortException);
+            _connectionWindow.Dispose();
+            _concurrentStreams.Dispose();
```

Comment:
Same question as earlier: why is it ok to revert the change here that passes _abortException into Dispose?

Reply:
Answered above